Gravatar/AvatarService.cs

```diff
@@ -0,0 +1,207 @@
+﻿using System;
+using System.Diagnostics;
+using System.Drawing;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+
+namespace Gravatar
+{
+    public interface IAvatarService
+    {
+        /// <summary>
+        /// Loads avatar either from the local cache or from the remote service.
+        /// </summary>
+        [NotNull]
+        [ItemNotNull]
+        Task<Image> GetAvatarAsync([NotNull] string email, int imageSize, [NotNull] string defaultImageType);
+
+        /// <summary>
+        /// Removes the avatar from the local cache.
+        /// </summary>
+        [NotNull]
+        Task DeleteAvatarAsync([NotNull] string email);
+
+        /// <summary>
+        /// Translates the serialised image type into the corresponding enum value.
+        /// </summary>
+        /// <param name="imageType">The serialised image type.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="imageType"/> is <c>null</c>.</exception>
+        DefaultImageType GetDefaultImageType([NotNull] string imageType);
+    }
+
+    public sealed class AvatarService : IAvatarService
+    {
+        private readonly IImageCache _cache;
+        private readonly IImageNameProvider _avatarImageNameProvider;
+
+        private const string GitHubPrivateEmailSuffix = "@users.noreply.github.com";
+
+        public AvatarService(IImageCache imageCache, IImageNameProvider avatarImageNameProvider = null)
+        {
+            _cache = imageCache;
+            _avatarImageNameProvider = avatarImageNameProvider ?? new AvatarImageNameProvider();
+        }
+
+        async Task<Image> IAvatarService.GetAvatarAsync(string email, int imageSize, string defaultImageType)
```

Comment:
The rationale here was to force consumption via the interface rather than the concrete type. If a concrete type has public members that aren't on the interface, then the interface isn't serving its purpose. I'm talking about types that fit the `class FooService : IFooService` pattern, which is more about abstraction for testability than polymorphism.

If the consensus is that this technique is not useful I can back it out. I mentioned it a while back in https://github.com/gitextensions/gitextensions/issues/4643#issuecomment-373725830 and no one said anything about it. It was in the middle of a wall of text, in fairness.